proto/pulumi/language.proto

```diff
@@ -23,290 +23,561 @@ package pulumirpc;
 
-    // RunPlugin executes a plugin program and returns its result asynchronously.
+    // `RunPlugin` is used to execute a program written in this host's language that implements a Pulumi
+    // [plugin](plugins). It it is plugins what [](pulumirpc.LanguageRuntime.Run) is to programs. Since a plugin is not
+    // expected to terminate until instructed/for a long time, this method returns a stream of
+    // [](pulumirpc.RunPluginResponse) messages containing information about standard error and output, as well as the
+    // exit code of the plugin when it does terminate.
     rpc RunPlugin(RunPluginRequest) returns (stream RunPluginResponse) {}
 
-    // GenerateProgram generates a given PCL program into a program for this language.
+    // `GenerateProgram` generates code in this host's language that implements the given [PCL](pcl) program. Unlike
+    // [](pulumirpc.LanguageRuntime.GenerateProject), this method *only* generates program code, and does not e.g.
+    // generate a `package.json` for a NodeJS project that details how to run that code.
+    // [](pulumirpc.LanguageRuntime.GenerateProject), this method underpins ["programgen"](programgen) and the main
+    // functionality powering `pulumi convert`.
     rpc GenerateProgram(GenerateProgramRequest) returns (GenerateProgramResponse) {}
 
-    // GenerateProject generates a given PCL program into a project for this language.
+    // `GenerateProject` generates code in this host's language that implements the given [PCL](pcl) program and wraps
+    // it in some language-specific notion of a "project", where a project is a buildable or runnable artifact. In this
+    // sense, `GenerateProject`'s output is a superset of that of [](pulumirpc.LanguageRuntime.GenerateProgram). For
+    // instance, when generating a NodeJS project, this method might generate a corresponding `package.json` file, as
+    // well as the relevant NodeJS program code. Along with [](pulumirpc.LanguageRuntime.GenerateProgram), this method
+    // underpins ["programgen"](programgen) and the main functionality powering `pulumi convert`.
     rpc GenerateProject(GenerateProjectRequest) returns (GenerateProjectResponse) {}
 
-    // GeneratePackage generates a given pulumi package into a package for this language.
+    // `GeneratePackage` generates code in this host's language that implements an [SDK](sdkgen) ("sdkgen") for the
+    // given Pulumi package, as specified by a [schema](schema).
     rpc GeneratePackage(GeneratePackageRequest) returns (GeneratePackageResponse) {}
 
-    // Pack packs a package into a language specific artifact.
+    // `Pack` accepts a request specifying a generated SDK package and packs it into a language-specific artifact. For
+    // instance, in the case of Java, it might produce a JAR file from a list of `.java` sources; in the case of NodeJS,
+    // a `.tgz` file might be produced from a list of `.js` sources; and so on. Presently, `Pack` is primarily used in
```

Comment:
I expect in the future it will be used as part of standard provider publishing workflows.